File: dotnet/Lexy.Poc/Lexy.Poc.Core/Compiler/LexyCompiler.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.IO;
using System.Linq;
using System.Reflection;
using Lexy.Poc.Core.Language;
using Lexy.Poc.Core.Parser;
using Lexy.Poc.Core.RunTime;
using Lexy.Poc.Core.Transcribe;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.Emit;
using Microsoft.Extensions.Logging;

namespace Lexy.Poc.Core.Compiler
{
    public class LexyCompiler : ILexyCompiler
    {
        private readonly IExecutionEnvironment environment;
        private readonly ICompilerContext compilerContext;

        public LexyCompiler(ICompilerContext compilerContext, IExecutionEnvironment environment)
        {
            this.compilerContext = compilerContext ?? throw new ArgumentNullException(nameof(compilerContext));
            this.environment = environment ?? throw new ArgumentNullException(nameof(environment));
        }

        public CompilerResult Compile(Components components, Function function)
        {
            if (components == null) throw new ArgumentNullException(nameof(components));
            if (function == null) throw new ArgumentNullException(nameof(function));

            var generateNodes = FunctionComponentAndDependencies(components, function);

            var code = GenerateCode(components, generateNodes);
            var assembly = CreateAssembly(code);

            environment.CreateExecutables(assembly);
            return environment.Result();
        }

        private static List<IRootComponent> FunctionComponentAndDependencies(Components components, Function function)
        {
            var generateNodes = new List<IRootComponent> { function };
            generateNodes.AddRange(function.GetDependencies(components));
            return generateNodes;
        }

        private Assembly CreateAssembly(string code)
        {
            var compilation = CreateCSharp(code);

            using var dllStream = new MemoryStream();
            using var pdbStream = new MemoryStream();

            var emitResult = compilation.Emit(dllStream, pdbStream);
            if (!emitResult.Success)
            {
                CompilationFailed(code, emitResult);
            }

            return Assembly.Load(dllStream.ToArray());
        }

        private static CSharpCompilation CreateCSharp(string code)
        {
            var syntaxTree = CSharpSyntaxTree.ParseText(code);
            var references = GetDllReferences();

            return CSharpCompilation.Create(
                $"{WriterCode.Namespace}.{Guid.NewGuid():D}",
                new[] { syntaxTree },
                references,
                new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary));
        }

        private void CompilationFailed(string code, EmitResult emitResult)
        {
            var compilationFailed = "Compilation failed: " +
                                    FormatCompilationErrors(emitResult.Diagnostics) +
                                    Environment.NewLine + "code: " + code;

            compilerContext.Logger.LogError(compilationFailed);
            throw new InvalidOperationException(compilationFailed);
        }

        private static List<MetadataReference> GetDllReferences()
        {
            var references = new List<MetadataReference>
            {
                MetadataReference.CreateFromFile(typeof(object).Assembly.Location),
                MetadataReference.CreateFromFile(typeof(FunctionResult).Assembly.Location)
            };

            Assembly.GetEntryAssembly().GetReferencedAssemblies()
                .ToList()
                .ForEach(reference =>
                    references.Add(MetadataReference.CreateFromFile(Assembly.Load(reference).Location)));
            return references;
        }

        private string GenerateCode(Components components, List<IRootComponent> generateComponents)
        {
            var classWriter = new ClassWriter();
            classWriter.WriteLine($"using System.Collections.Generic;");
            classWriter.WriteLine($"using {typeof(IExecutionContext).Namespace};");
            classWriter.WriteLine();
            classWriter.OpenScope($"namespace {WriterCode.Namespace}");

            foreach (var component in generateComponents)
            {
                var writer = GetWriter(component);
                var generatedType = writer.CreateCode(classWriter, component, components);

                environment.AddType(generatedType);
            }

            classWriter.CloseScope();

            var code = classWriter.ToString();
            compilerContext.Logger.LogDebug("Compile code: " + Environment.NewLine + code);
            return code;
        }

        private static IRootTokenWriter GetWriter(IRootComponent rootComponent)
        {
            return rootComponent switch
            {
                Function _ => new FunctionWriter(),
                EnumDefinition _ => new EnumWriter(),
                Table _ => new TableWriter(),
                Scenario _ => null,
                _ => throw new InvalidOperationException("No writer defined: " + rootComponent.GetType())
            };
        }

        private static string FormatCompilationErrors(ImmutableArray<Diagnostic> emitResult)
        {
            var stringWriter = new StringWriter();
            foreach (var diagnostic in emitResult)
            {
                stringWriter.WriteLine($"  {diagnostic}");
            }
            return stringWriter.ToString();
        }
    }
}

File: typescipt/src/compiler/lexyCompiler.ts


export class LexyCompiler extends ILexyCompiler {
   private readonly ILogger<LexyCompiler> logger;
   private readonly IExecutionEnvironment environment;

   constructor(logger: ILogger<LexyCompiler>, environment: IExecutionEnvironment) {
     this.logger = logger ?? throw new Error(nameof(logger));
     this.environment = environment ?? throw new Error(nameof(environment));
   }

   public compile(nodes: Array<IRootNode>): CompilerResult {
     if (nodes == null) throw new Error(nameof(nodes));

     try {
       let syntaxNode = GenerateSyntaxNode(nodes);
       let assembly = CreateAssembly(syntaxNode);

       environment.CreateExecutables(assembly);
       return environment.Result();
     }
     catch (Exception e) {
       logger.LogError(e, `Exception occured during compilation`);
       throw;
     }
   }

   private createAssembly(syntax: SyntaxNode): Assembly {
     let compilation = CreateSyntaxTree(syntax);

     string fullString = null;
     if (logger.IsEnabled(LogLevel.Debug)) {
       fullString = syntax.ToFullString();
       logger.LogDebug(fullString);
     }

     using let dllStream = new MemoryStream();
     using let pdbStream = new MemoryStream();

     let emitResult = compilation.Emit(dllStream, pdbStream);
     if (!emitResult.Success) CompilationFailed(fullString ?? syntax.ToFullString(), emitResult);

     return Assembly.Load(dllStream.ToArray());
   }

   private static createSyntaxTree(root: SyntaxNode): CSharpCompilation {
     let syntaxTree = SyntaxTree(root);
     let references = GetDllReferences();

     return CSharpCompilation.Create(
       $`{LexyCodeConstants.Namespace}.{Guid.NewGuid():D}`,
       new[] { syntaxTree },
       references,
       new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary));
   }

   private compilationFailed(code: string, emitResult: EmitResult): void {
     let compilationFailed = $`Compilation failed: {FormatCompilationErrors(emitResult.Diagnostics)}`;

     logger.LogError(compilationFailed);

     throw new Error($`{compilationFailed}{Environment.NewLine}code: {code}`);
   }

   private static getDllReferences(): Array<MetadataReference> {
     let references = new Array<MetadataReference> {
       MetadataReference.CreateFromFile(typeof(object).Assembly.Location),
       MetadataReference.CreateFromFile(typeof(FunctionResult).Assembly.Location)
     };

     Assembly.GetEntryAssembly().GetReferencedAssemblies()
       .ToList()
       .ForEach(reference =>
         references.Add(MetadataReference.CreateFromFile(Assembly.Load(reference).Location)));

     return references;
   }

   private generateSyntaxNode(generateNodes: Array<IRootNode>): SyntaxNode {
     let root = GenerateCompilationUnitS(generateNodes);

     return root.NormalizeWhitespace();
   }

   private generateCompilationUnitS(generateNodes: Array<IRootNode>): CompilationUnitSyntax {
     let members = generateNodes
       .Select(GenerateMember)
       .ToList();

     let namespaceDeclaration = NamespaceDeclaration(IdentifierName(LexyCodeConstants.Namespace))
       .WithMembers(List(members));

     let root = CompilationUnit()
       .WithUsings(List(
         new[] {
           Using(`System`),
           Using(`System.Collections.Generic`),
           Using(typeof(IExecutionContext).Namespace)
         }))
       .WithMembers(SingletonArray<MemberDeclarationSyntax>(namespaceDeclaration));
     return root;
   }

   private generateMember(node: IRootNode): MemberDeclarationSyntax {
     let writer = CSharpCode.GetWriter(node);

     let generatedType = writer.CreateCode(node);

     environment.AddType(generatedType);

     return generatedType.Syntax;
   }

   private static using(ns: string): UsingDirectiveSyntax {
     return UsingDirective(ParseName(ns));
   }

   private static formatCompilationErrors(emitResult: ImmutableArray<Diagnostic>): string {
     let stringWriter = new StringWriter();
     foreach (let diagnostic in emitResult) stringWriter.WriteLine($` {diagnostic}`);

     return stringWriter.ToString();
   }
}
